mutiny-core/src/federation.rs

```diff
@@ -0,0 +1,216 @@
+        federation_code: InviteCode,
+        xprivkey: ExtendedPrivKey,
+        storage: S,
+        network: Network,
+        logger: Arc<MutinyLogger>,
+        stop: Arc<AtomicBool>,
+    ) -> Result<Self, MutinyError> {
+        log_info!(logger, "initializing a new federation client: {uuid}");
+
+        // a list of components that need to be stopped and whether or not they are stopped
+        // TODO remove this if we end up not needing to stop things
+        let stopped_components = Arc::new(RwLock::new(vec![]));
+
+        log_info!(logger, "Joining federation {}", federation_code);
+
+        let federation_info = FederationInfo::from_invite_code(federation_code.clone()).await?;
+
+        let mut client_builder = fedimint_client::Client::builder();
+        client_builder.with_module(WalletClientInit(None));
+        client_builder.with_module(MintClientInit);
+        client_builder.with_module(LightningClientInit);
+        client_builder.with_database(MemDatabase::new().into()); // TODO not in memory
+        client_builder.with_primary_module(1);
+        client_builder
+            .with_federation_info(FederationInfo::from_invite_code(federation_code.clone()).await?);
+
+        let secret = create_federation_secret(xprivkey, federation_index, network)?;
+
+        let fedimint_client = client_builder
+            .build(get_default_client_secret(
+                &secret,
+                &federation_info.federation_id(),
+            ))
+            .await?;
+
+        Ok(FederationClient {
+            _uuid: uuid,
+            federation_index: federation_index.clone(),
+            federation_code,
+            fedimint_client,
+            stopped_components,
+            storage,
+            network,
+            logger,
+            stop,
+        })
+    }
+
+    pub(crate) fn federation_index(&self) -> FederationIndex {
+        FederationIndex {
+            child_index: self.federation_index.child_index,
+            federation_code: self.federation_code.clone(),
+            archived: self.federation_index.archived,
+        }
+    }
+}
+
+// A federation private key will be derived from `m/1'/N'/X'`, where:
+// - N is the network type
+// - X is the index of a specific federation.
+//
+// Federation will derive further keys from there.
+fn create_federation_secret(
+    xprivkey: ExtendedPrivKey,
+    federation_index: &FederationIndex,
+    network: Network,
+) -> Result<DerivableSecret, MutinyError> {
+    let context = Secp256k1::new();
+    let xpriv = xprivkey.derive_priv(
+        &context,
+        &DerivationPath::from(vec![
+            ChildNumber::from_hardened_idx(1)?,
+            ChildNumber::from_hardened_idx(coin_type_from_network(network))?,
+            ChildNumber::from_hardened_idx(federation_index.child_index)?,
+        ]),
+    )?;
+    let secret =
+        DerivableSecret::new_root(&xpriv.private_key.secret_bytes(), FEDIMINT_CLIENT_NONCE);
+    Ok(secret)
+}
```

Comment:
Was thinking about this more, do we need the `federation_index`? Don't see why we need different keys for each federation. I don't think there is a privacy implication since it is all ecash, could be wrong though.

If we got rid of it it'd make things easier where if the user has their seed they don't also need to know the index for each federation.

Reply:
This is a valid point. I can remove it and just use the master seed but be aware, both fedimint and our LN nodes will be using `m/0'/` as the first part of the prefix. I'm afraid of potential clashes.

Reply:
These secondary indexes should be hardcoded and stored somewhere.